src/helpers.c

```diff
@@ -191,88 +191,80 @@ int system_argv(char *const argv[])
 	return status;
 }
 
-int system_argv_fd(char *const argv[], int newstdin, int newstdout, int newstderr)
+int system_argv_pipe(char *const lhscmd[], char *const rhscmd[])
 {
-	int child_exit_status;
-	pid_t pid;
-	int status = -1;
-
-	pid = fork();
-
-	if (pid == 0) { /* child */
-		if (newstdin >= 0) {
-			if (dup2(newstdin, STDIN_FILENO) == -1) {
-				LOG(NULL, "Could not redirect stdin", "");
-				assert(0);
-			}
-			close(newstdin);
+	pid_t childpid = fork();
+	if (childpid == -1) {
+		LOG(NULL, "Failed to create child process", "");
+		return -1;
+	} else if (childpid == 0) {
+		int pipefd[2];
+		if (pipe(pipefd) == -1) {
+			LOG(NULL, "Failed to create a pipe", "");
+			assert(0);
 		}
-		if (newstdout >= 0) {
-			if (dup2(newstdout, STDOUT_FILENO) == -1) {
+
+		pid_t lhspid = fork();
+		if (lhspid == -1) {
+			LOG(NULL, "Failed to fork LHS process", "");
+			return -1;
+		} else if (lhspid == 0) {
+			if (dup2(pipefd[1], STDOUT_FILENO) == -1) {
 				LOG(NULL, "Could not redirect stdout", "");
 				assert(0);
 			}
-			close(newstdout);
-		}
-		if (newstderr >= 0) {
-			if (dup2(newstderr, STDERR_FILENO) == -1) {
-				LOG(NULL, "Could not redirect stderr", "");
+			execvp(*lhscmd, lhscmd);
+			LOG(NULL, "LHS command failed", "");
+			assert(0);
+		} else {
+			pid_t rhspid = fork();
+			if (rhspid == -1) {
+				LOG(NULL, "Failed to fork RHS process", "");
+				return -1;
+			} else if (rhspid == 0) {
+				if (dup2(pipefd[0], STDIN_FILENO) == -1) {
+					LOG(NULL, "Could not redirect stdin", "");
+					assert(0);
+				}
+				execvp(*rhscmd, rhscmd);
+				LOG(NULL, "LHS command failed", "");
 				assert(0);
+			} else {
+				int lhsresult = wait_process_terminate(lhspid);
```

Comment:
It is **critical** that the 2 close calls are done before the 2 wait_process_terminate calls, to avoid deadlocks.